### ISSUE TITLE
ISTFT head use center padding but the same padding

### ISSUE THREAD

**Issue by nukes** [charactr-platform/vocos#18]
Hi
 this is a great work. I noticed that the ISTFT head does not use the same padding that you implemented, but uses the center padding which is provided by torch api according your config file.
for neural vocoding, the spectrum frames and samples should be time aligned(num_frames* hop_len = samples), why did not you use the same padding you implemented for that property?

**Comment by hubertsiuzdak** [comment]
This setting needs to be consistent with the feature extractor. I guess you specifically mean the [vocos.yaml ](https://github.com/charactr-platform/vocos/blob/86db50a025ecaddf864366d5dbfc83554ceedc60/configs/vocos.yaml) config file that takes mel-spectrograms as inputs. Note that input mel-spectrograms are "center" padded, so to compensate for this we can use `torch.istft` with `center=True` which trims the corresponding samples.

However, for features like EnCodec tokens that are not padded in that specific way, using `torch.istft` with `center=True` would trim too many samples. In the [vocos-encodec.yaml](https://github.com/charactr-platform/vocos/blob/86db50a025ecaddf864366d5dbfc83554ceedc60/configs/vocos-encodec.yaml) config file, you'll find `padding: same` in the `ISTFTHead`.

It would certainly be simpler if we could use `torch.istft` with `center=False` and slice the output audio. However, PyTorch does not allow this (for specific windows) due to how the NOLA (Nonzero Overlap Add) is checked. You might want to check out this issue: https://github.com/pytorch/pytorch/issues/91309

Hope it helps!

**Comment by HeJinLing** [comment]
Hi, i noticed that the parameter of n_fft in head is 1280, why did not you use 1024?

**Comment by hubertsiuzdak** [comment]
@HeJinLing 

The key parameter here is `hop_length`. It should align with the resolution of your input features. Since EnCodec tokens are downsampled by a factor of 320, we've set the `hop_length` to 320.

Now, when you use a Hann window in the iSTFT, it's common to have a 75% overlap. This means our `window_len` should be four times the `hop_length`. That's why we set `window_len` (and `n_fft`) to 1280. 

If you want to dive a bit deeper, I'd recommend looking into the constant overlap-add (COLA) constraint, there's a helpful discussion on this topic here: https://dsp.stackexchange.com/a/33615.

**Comment by zaptrem** [comment]
> @HeJinLing
> 
> The key parameter here is `hop_length`. It should align with the resolution of your input features. Since EnCodec tokens are downsampled by a factor of 320, we've set the `hop_length` to 320.
> 
> Now, when you use a Hann window in the iSTFT, it's common to have a 75% overlap. This means our `window_len` should be four times the `hop_length`. That's why we set `window_len` (and `n_fft`) to 1280.
> 
> If you want to dive a bit deeper, I'd recommend looking into the constant overlap-add (COLA) constraint, there's a helpful discussion on this topic here: https://dsp.stackexchange.com/a/33615.

On this note, @hubertsiuzdak, would architectural changes be necessary if `hop_length` is pushed to the extreme? Over the past few days I've been trying to train one with `hop_length = 4096` and `n_ftt = 8192` to fit my latent signal from another project. Unfortunately, every set of hparams I've tried has resulted in the discriminators winning. Training a large (~200m parameter) model with larger discriminators has gotten the furthest thus far, but seems to eventually succumb to the feature matching loss. My config is as follows:

```
backbone:
  class_path: vocos.models.VocosBackbone
  init_args:
    input_channels: 64
    dim: 2560
    intermediate_dim: 5120
    num_layers: 8
    # adanorm_num_embeddings: None  # Only one bandwidth

head:
  class_path: vocos.heads.ISTFTHead
  init_args:
    dim: 2560
    n_fft: 8192 
    hop_length: 4096 
    padding: same
    stereo: true # just another linear layer that does dim -> (n_fft + 2) * 2 and runs each through the stft to get a stereo signal.
```

I also tend to get a large click at the start of samples as training continues.

Thanks for any wisdom you can provide!

**Comment by jishengpeng** [comment]
I have been experimenting with modifying the **hop_length** and **n_fft** parameters in the Vocos framework, but the results have been suboptimal.

Specifically, in the encoder section, I am utilizing the Encodec with a sampling rate of 24000 Hz. I attempted to compress the codec to achieve frame rates of 25 frames per second (fps) and 100 fps, as opposed to the original 75 fps of the Encodec. The specific configurations are as follows:

For 100 fps: **hop_length**=240, **n_fft**=960
For 25 fps: **hop_length**=960, **n_fft**=3840

Unfortunately, both configurations have yielded very poor results. Could you provide insights into the potential reasons for this issue

**Comment by zaptrem** [comment]
> I have been experimenting with modifying the **hop_length** and **n_fft** parameters in the Vocos framework, but the results have been suboptimal.
> 
> Specifically, in the encoder section, I am utilizing the Encodec with a sampling rate of 24000 Hz. I attempted to compress the codec to achieve frame rates of 25 frames per second (fps) and 100 fps, as opposed to the original 75 fps of the Encodec. The specific configurations are as follows:
> 
> For 100 fps: **hop_length**=240, **n_fft**=960 For 25 fps: **hop_length**=960, **n_fft**=3840
> 
> Unfortunately, both configurations have yielded very poor results. Could you provide insights into the potential reasons for this issue

You should try to train your own Encodec model with your specific token requirements before trying to train a Vocos on it.

**Comment by jishengpeng** [comment]
> > I have been experimenting with modifying the **hop_length** and **n_fft** parameters in the Vocos framework, but the results have been suboptimal.
> > Specifically, in the encoder section, I am utilizing the Encodec with a sampling rate of 24000 Hz. I attempted to compress the codec to achieve frame rates of 25 frames per second (fps) and 100 fps, as opposed to the original 75 fps of the Encodec. The specific configurations are as follows:
> > For 100 fps: **hop_length**=240, **n_fft**=960 For 25 fps: **hop_length**=960, **n_fft**=3840
> > Unfortunately, both configurations have yielded very poor results. Could you provide insights into the potential reasons for this issue
> 
> You should try to train your own Encodec model with your specific token requirements before trying to train a Vocos on it.

Thank you for your response. However, I have indeed retrained the encoder and decoder together in an end-to-end manner. Nevertheless, the aforementioned configuration has led to subpar results.

**Comment by Ted-Picovoice** [comment]
> This setting needs to be consistent with the feature extractor. I guess you specifically mean the [vocos.yaml ](https://github.com/charactr-platform/vocos/blob/86db50a025ecaddf864366d5dbfc83554ceedc60/configs/vocos.yaml) config file that takes mel-spectrograms as inputs. Note that input mel-spectrograms are "center" padded, so to compensate for this we can use `torch.istft` with `center=True` which trims the corresponding samples.
> 
> However, for features like EnCodec tokens that are not padded in that specific way, using `torch.istft` with `center=True` would trim too many samples. In the [vocos-encodec.yaml](https://github.com/charactr-platform/vocos/blob/86db50a025ecaddf864366d5dbfc83554ceedc60/configs/vocos-encodec.yaml) config file, you'll find `padding: same` in the `ISTFTHead`.
> 
> It would certainly be simpler if we could use `torch.istft` with `center=False` and slice the output audio. However, PyTorch does not allow this (for specific windows) due to how the NOLA (Nonzero Overlap Add) is checked. You might want to check out this issue: [pytorch/pytorch#91309](https://github.com/pytorch/pytorch/issues/91309)
> 
> Hope it helps!

Hi!

I'd wish to ask is the only difference between "same" padding and "center" padding in the case with STFT and ISTFT is that "center" padding pads `n_fft // 2` on both sides whereas "same" padding pads `(n_fft - hop_length) // 2 ` on both sides?

Why would one prefer over another? I just don't see a big difference between the two.

Thanks!
